### PR TITLE
Tile Grid Revamp

### PR DESCRIPTION
**Pull request by toshokanneko**
- Consistent grid (no empty spots or incomplete lines....unless you reach the bottom of a category or the pageSize is set to a weird value like in the German Music section)
- No more shifting when the livestreams are added to the grid on the homepage
- Custom Premium+ tiles when ads are blocked
- Pulse animation for placeholder tiles is back
- Better integrated ads (visually)
- Minor design/UI improvements

### CODE REVIEW

File: ui/scss/component/_ads.scss [hunk 1]
```diff
@@ -33,10 +33,17 @@
     aspect-ratio: 16 / 9;
     $minWidth: calc(var(--file-list-thumbnail-width) * 0.8);
     min-width: $minWidth;
+    background-color: #283263;
+    background-image: url('https://odysee.com/public/img/astronaut_n_friends.png');
+    background-size: 100%;
+    border-radius: var(--border-radius);
+    box-shadow: 0px 0px 0px 1px rgba(var(--color-primary-dynamic), 0.1) inset;
```
Comment (infinite-persistence) [inline ui/scss/component/_ads.scss:40]: webstorm says it''s redundant
(which by the way, we should probably add a css linter in the near future?)
image

**Review by infinite-persistence — CHANGES_REQUESTED** [review]

File: ui/component/claimPreviewTile/view.jsx [hunk 2]
```diff
@@ -55,6 +55,8 @@ type Props = {
   isLivestreamActive: boolean,
   livestreamViewerCount: ?number,
   swipeLayout: boolean,
+  isHidden: (string) => void,
+  pulse: boolean,
```
[the remark this answers is not in the document]
Comment (toshokanneko) [reply ui/component/claimPreviewTile/view.jsx:59]: Nope, that's how it used to be before we've decided to disable the animation. If we add it to all placeholders the homepage starts to lag again because the animation would play on every single tile for each category.

File: ui/page/home/view.jsx [hunk 3]
```diff
@@ -169,6 +169,8 @@ function HomePage(props: Props) {
           }
         }
         forceShowReposts={id !== 'FOLLOWING'}
+        loading={id === 'FOLLOWING' ? fetchingActiveLivestreams : false}
```
[the remark this answers is not in the document]
Comment (toshokanneko) [reply ui/page/home/view.jsx:172]: This change is "necessary" in order to prevent the shift in the grid when the livestreams get added. So yeah, it's specific to the livestreams. That's currently the most annoying thing on the homepage. Blocklists etc. are also an issue, but those get handled one level deeper for each tile individually. We have to come up with a better solution for those.

File: ui/component/scheduledStreams/view.jsx [hunk 4]
```diff
@@ -21,8 +21,10 @@ type Props = {
   // --- perform ---
   setClientSetting: (string, boolean | string | number, boolean) => void,
   doShowSnackBar: (string) => void,
+  setLoaded: (boolean) => void,
```
[the remark this answers is not in the document]
Comment (toshokanneko) [reply ui/component/scheduledStreams/view.jsx:24]: Yeah, the idea was to prevent the shift when scheduled livestreams get added to the homepage, but Tom has mentioned that these calls are about to change in the near future and that this isn't necessary. Will clean it up.

File: ui/component/claimTilesDiscover/view.jsx [hunk 5]
```diff
@@ -200,30 +233,39 @@ function ClaimTilesDiscover(props: Props) {
...
+                return (
+                  <React.Fragment key={uri}>
+                    <PremiumPlus />
+                  </React.Fragment>
+                );
+              } else if (inj) {
+                return <React.Fragment key={uri}>{inj}</React.Fragment>;
+              } else {
+                return (
+                  <React.Fragment key={uri}>
```
[the remark this answers is not in the document]
Comment (toshokanneko) [reply ui/component/claimTilesDiscover/view.jsx:250]: Okay, makes sense on the channel pages, but injections are the reason why the current grid is broken. Who came up with this idea? It's always gonna replace an element on the individual level instead of the grid level which makes it impossible to adjust the array without constant communication between the parent and the child.

File: ui/scss/component/_ads.scss (hunk 1 above)
Comment (toshokanneko) [reply ui/scss/component/_ads.scss:40]: It is redundant, but I usually only remove the units when all the values are 0. In this case there's still 1px for the bottom left. I can remove it tho...

File: ui/component/claimTilesDiscover/view.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (infinite-persistence) [reply ui/component/claimTilesDiscover/view.jsx:250]: Agree. My second bullet point could be a misunderstanding, so can ignore.  Thought process was:
- "replace" was the original implementation in Homepage, then was asked to do "insert" because a user complaint about a missing tile from creator in the homepage, which I didn't understand since it's a truncated list and there is a "Show more" button to see everything.  Then, I thought the next request is to continue with "insert" but fill remaining space, which I didn't know how since sometimes there is no more data to fill with.

File: ui/scss/component/_ads.scss (hunk 1 above)
Comment (infinite-persistence) [reply ui/scss/component/_ads.scss:40]: No worries, was just passing along info.
We follow https://github.com/airbnb/css style guide per https://github.com/OdyseeTeam/odysee-frontend/blob/master/CONTRIBUTING.md, and I always thought the rule was part of it, but seems like not.

File: ui/component/claimTilesDiscover/view.jsx [hunk 6]
```diff
@@ -67,6 +70,9 @@ type Props = {
   doFetchUserMemberships: (claimIdCsv: string) => void,
   doResolveClaimIds: (Array<string>) => Promise<any>,
   doResolveUris: (Array<string>, boolean) => Promise<any>,
+  loading: boolean,
+  hasPremiumPlus: ?boolean,
+  adBlockerFound: boolean,
```
[the remark this answers is not in the document]
Comment (toshokanneko) [reply ui/component/claimTilesDiscover/view.jsx:75]: For some reason I've never even noticed these markers before. I'll use them from now on, thx.

File: ui/component/claimTilesDiscover/view.jsx [hunk 7]
```diff
@@ -200,30 +233,39 @@ function ClaimTilesDiscover(props: Props) {
...
     <ul ref={listRef} className="claim-grid">
-      {finalUris && finalUris.length
+      {!loading && finalUris && finalUris.length
         ? finalUris.map((uri, i) => {
             if (uri) {
               const inj = getInjectedItem(i);
-              return (
-                <React.Fragment key={uri}>
-                  {inj && inj}
-                  {(!inj || !injectedItem || !injectedItem.replace) && (
```
[the remark this answers is not in the document]
Comment (toshokanneko) [reply ui/component/claimTilesDiscover/view.jsx:210]: Okay, I've completely removed the replace option. Now I just have to find away to make the injection work without skipping any tiles...

File: .env.defaults [hunk 8]
```diff
@@ -57,7 +57,7 @@ SHOW_ADS=true
 ENABLE_MATURE=false
 ENABLE_UI_NOTIFICATIONS=true
 #SHOW_TAGS_INTRO=false
-USE_LOCAL_HOMEPAGE_DATA=true
+USE_LOCAL_HOMEPAGE_DATA=false
```
Comment (infinite-persistence) [inline .env.defaults:60]: I assume this is temporary and not needed for this PR?

Aside: most likely you already now, but you can create a blank `web/.env` file locally, and add `USE_LOCAL_HOMEPAGE_DATA=false` (and full rebuild, can't hot reload).
This saves the need to commit and revert.  Also works for dev instances (need to ssh in).

File: .gitignore [hunk 9]
```diff
@@ -37,4 +37,5 @@ package-lock.json
 .env.ody
 .env.desktop
 .env.lbrytv
+.env.defaults
```
Comment (infinite-persistence) [inline .gitignore:40]: Defaults is needed in production build.  Use local override method mentioned above.

File: ui/component/claimList/index.js [hunk 10]
```diff
@@ -3,9 +3,13 @@ import ClaimList from './view';
 import * as SETTINGS from 'constants/settings';
 
 import { selectClientSetting } from 'redux/selectors/settings';
+import { selectOdyseeMembershipIsPremiumPlus } from 'redux/selectors/user';
+import { selectAdBlockerFound } from 'redux/selectors/app';
 
 const select = (state) => ({
   searchInLanguage: selectClientSetting(state, SETTINGS.SEARCH_IN_LANGUAGE),
+  hasPremiumPlus: selectOdyseeMembershipIsPremiumPlus(state),
+  adBlockerFound: selectAdBlockerFound(state),
```
Comment (infinite-persistence) [inline ui/component/claimList/index.js:12]: I think can remove now that the PremiumTile logic is at the caller side.
Clean up jsx side too.

File: ui/component/claimList/view.jsx [hunk 11]
```diff
@@ -228,21 +228,34 @@ export default function ClaimList(props: Props) {
...
+                    <ClaimPreviewTile
+                      uri={uri}
+                      showHiddenByUser={showHiddenByUser}
+                      showUnresolvedClaims={showUnresolvedClaims}
+                      properties={renderProperties}
+                      collectionId={collectionId}
+                      fypId={fypId}
+                      showNoSourceClaims={showNoSourceClaims}
+                      swipeLayout={swipeLayout}
+                      isHidden={onHidden}
```
Comment (infinite-persistence) [inline ui/component/claimList/view.jsx:252]: `isHidden` parameter typo

File: ui/component/claimTilesDiscover/index.js [hunk 12]
```diff
@@ -42,6 +43,7 @@ const select = (state, props) => {
     claimsByUri: selectClaimsByUri(state),
     claimsById: selectById(state),
     fetchingClaimSearch: selectFetchingClaimSearchByQuery(state)[searchKey],
+    hasPremiumPlus: selectOdyseeMembershipIsPremiumPlus(state),
```
Comment (infinite-persistence) [inline ui/component/claimTilesDiscover/index.js:46]: no longer used

File: ui/component/claimTilesDiscover/view.jsx [hunk 13]
```diff
@@ -61,6 +62,8 @@ type Props = {
   showNsfw: boolean,
   hideReposts: boolean,
   optionsStringified: string,
+  hasPremiumPlus: ?boolean,
+  adBlockerFound: boolean,
```
Comment (infinite-persistence) [inline ui/component/claimTilesDiscover/view.jsx:66]: No longer necessary?

File: ui/page/home/view.jsx [hunk 14]
```diff
@@ -238,6 +248,7 @@ function HomePage(props: Props) {
               tileLayout
               liveUris={getLivestreamUris(activeLivestreams, channelIds)}
               limitClaimsPerChannel={2}
+              // setLoaded={setLoaded}
```
Comment (infinite-persistence) [inline ui/page/home/view.jsx:251]: Remove

**Review by infinite-persistence — CHANGES_REQUESTED** [review]

File: ui/page/home/view.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (infinite-persistence) [reply ui/page/home/view.jsx:180]: Ah ok.  The original behavior was to not mount the list at all until livestreams are fetched, but that caused the homepage to be empty for longer than necessary.  This at least places the placeholder tiles :+1:

[the remark this answers is not in the document]
Comment (toshokanneko) [reply ui/page/home/view.jsx:175]: Jep and the site loads MUCH faster than it used to. At least in most regions.

File: ui/component/claimList/view.jsx [hunk 15]
```diff
@@ -33,7 +33,7 @@ type Props = {
   defaultSort?: boolean,
   onScrollBottom?: (any) => void,
   page?: number,
-  pageSize?: number,
+  pageSize: number,
```
Comment (infinite-persistence) [inline ui/component/claimList/view.jsx:36]: `pageSize` should remain optional, I believe. There are lots of `<ClaimList>` instances without `pageSize` and it works.

File: ui/component/claimList/view.jsx [hunk 16]
```diff
@@ -61,6 +61,9 @@ type Props = {
   unavailableUris?: Array<string>,
   showMemberBadge?: boolean,
   inWatchHistory?: boolean,
+  onHidden: string,
+  hasPremiumPlus: ?boolean,
+  adBlockerFound: boolean,
```
Comment (infinite-persistence) [inline ui/component/claimList/view.jsx:66]: Remove `hasPremiumPlus` and `adBlockerFound`
Remove `hasPremiumPlus` usage down below as the logic is now at the caller side.

File: ui/component/claimTilesDiscover/view.jsx [hunk 17]
```diff
@@ -61,6 +62,7 @@ type Props = {
   showNsfw: boolean,
   hideReposts: boolean,
   optionsStringified: string,
+  adBlockerFound: boolean,
```
Comment (infinite-persistence) [inline ui/component/claimTilesDiscover/view.jsx:65]: :x:

File: ui/scss/component/_ads.scss [hunk 18]
```diff
@@ -203,8 +299,60 @@
...
+
 .ob-widget-items-container {
   padding-left: var(--spacing-xs);
   padding-right: var(--spacing-xs);
-  border-radius: var(--border-radius);
+  border-radius: var(--border-radius) 0 0 0;
+}
+
+// ****************************************************************************
+// Neko Patch
```
Comment (infinite-persistence) [inline ui/scss/component/_ads.scss:346]: Corner-case

image

**Review by infinite-persistence — COMMENTED** [review]
Didn't test it live -- relying on Miko's test which seems to be positive.
Did we test mobile?  I know it's not directly affected, but never know.

A few minor cleanups and then good to go.

File: ui/component/claimList/view.jsx (hunk 15 above)
Comment (infinite-persistence) [reply ui/component/claimList/view.jsx:36]: The usual fix is to make the code handle the undefined case (skip the loop or something), instead of the other way around of changing the definition to make it pass.  The warning is to remind us of all possible code paths.

Example existing code does things only when page and pageSize is defined `if (page && pageSize && onScrollBottom) {`
